cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1720,7 +1720,27 @@ class S3FileSystem::Impl : public std::enable_shared_from_this<S3FileSystem::Imp
   // a non-empty "directory".  This is a Minio-specific quirk, but we need
   // to handle it for unit testing.
 
-  Status IsEmptyDirectory(const std::string& bucket, const std::string& key, bool* out) {
+  // If this method is called after HEAD on "bucket/key" already returned a 404,
+  // can pass the given outcome to spare a spurious HEAD call.
+  Result<bool> IsEmptyDirectory(
+      const std::string& bucket, const std::string& key,
+      const S3Model::HeadObjectOutcome* previous_outcome = nullptr) {
+    if (previous_outcome) {
+      // Fetch the backend from the previous error
+      DCHECK(!previous_outcome->IsSuccess());
+      if (!backend_) {
+        SaveBackend(previous_outcome->GetError());
+        DCHECK(backend_);
+      }
+      if (backend_ != S3Backend::Minio) {
+        // HEAD already returned a 404, nothing more to do
+        return false;
+      }
+    }
+
+    // We come here in one of two situations:
+    // - we don't know the backend and there is no previous outcome
```

Comment:
```suggestion
    // - there is no previous outcome
```

Reply:
The backend is cached on the filesystem, so I think your formulation is less accurate.

Reply:
The working sounds like "we don't know the backend" (A) and "there is no previous outcome" (B) both have to be true (A & B).  But it seems like we would get here on (!A & B) as well.

Reply:
If we know the backend, we only come here if the backend is Minio, and that is the second situation :-) Perhaps the wording is confusing?

Reply:
No, the wording is clear.  My understanding was not.  I see what you are saying now.  I hadn't realized !A => minio.